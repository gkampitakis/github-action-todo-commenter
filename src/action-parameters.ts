import { getInput } from '@actions/core';
import { Context } from '@actions/github/lib/context';

import { GetActionParams, GetInputParams } from './types';

export function getInputs(): GetInputParams {
  const tags = getInput('tags') || 'TODO:,FIXME:,BUG:';
  const reviewMsg = getInput('review-message');
  const ignoreFilesPattern = getInput('ignore-pattern');
  const token = getInput('github-token', { required: true });
  const commentTitle = getInput('comment-title') || 'Todo Commenter';

  return {
    tags: tags.split(','),
    reviewMsg,
    token,
    ignoreFilesPattern,
    commentTitle
  };
}

export function getActionParameters(ctx: Context): GetActionParams {
  const {
    actor,
    repo: { owner, repo },
    eventName,
    payload: { pull_request }
  } = ctx;
  const base = pull_request?.base?.ref;
  const head = pull_request?.head?.ref;

  console.log(eventName);

  if (eventName !== 'pull_request') {
    throw new Error('Action only supports pull requests');
  }

  if (!pull_request?.number) {
    throw new Error('Action cannot identify pull request number');
  }

  return {
    actor,
    owner,
    repo,
    base,
    head,
    prNumber: pull_request?.number
  };
}
