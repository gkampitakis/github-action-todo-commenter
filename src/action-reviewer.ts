import { info } from '@actions/core';
import { ActionReviewerOptions, Octokit } from './types';

export class ActionReviewer {
  private octokit: Octokit;
  private options: ActionReviewerOptions;

  constructor(octokit: Octokit, options: ActionReviewerOptions) {
    this.octokit = octokit;
    this.options = options;
  }

  public async createReview(body: string) {
    const { id, body: oldReviewBody } = await this.reviewExists();

    if (oldReviewBody === body) {
      info('No new changes detected since ');
      return;
    }

    if (id) {
      await this.deleteReview(id);
    }

    await this.octokit.rest.issues.createComment({
      owner: this.options.owner,
      repo: this.options.repo,
      issue_number: this.options.prNumber,
      body
    });

    return;
  }

  public async singleCommentReview(
    body: string,
    { line, path }: { line: number; path: string }
  ) {
    return this.octokit.rest.pulls.createReviewComment({
      owner: this.options.owner,
      repo: this.options.repo,
      pull_number: this.options.prNumber,
      commit_id: this.options.commitId,
      position: line,
      body,
      path,
      line
    });
  }
// test
  public async deleteReview(id: number) {
    return this.octokit.rest.issues.deleteComment({
      owner: this.options.owner,
      repo: this.options.repo,
      comment_id: id
    });
  }

  public async reviewExists(): Promise<{ id?: number; body?: string }> {
    const { data: allReviews } = await this.octokit.rest.issues.listComments({
      owner: this.options.owner,
      repo: this.options.repo,
      issue_number: this.options.prNumber
    });

    if (allReviews.length === 0) {
      return {};
    }

    const review = allReviews.find(item => {
      return (
        item.user?.login === 'github-actions[bot]' &&
        item.body?.startsWith(this.options.commentIdentifier)
      );
    });

    if (!review || !review.body) {
      return {};
    }

    // This removes the first line containing the comment identifier
    let index = review.body.match(/[\r\n]+/)?.index;
    index = index ? index + 1 : 0;

    return { id: review?.id, body: review.body.substr(index) };
  }
}
