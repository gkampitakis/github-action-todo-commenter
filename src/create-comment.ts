import { getOctokit } from '@actions/github';
type CreateCommentParams = {
  token: string;
  owner: string;
  repo: string;
  prNumber: number;
};

export async function createComment(
  params: CreateCommentParams,
  body: string,
  block: boolean
) {
  const { token, owner, repo, prNumber } = params;
  const octokit = getOctokit(token);

  const { data: allComments } = await octokit.rest.issues.listComments({
    owner,
    repo,
    issue_number: prNumber
  });

  const comment = allComments.find(item => {
    return (
      item.user?.login === 'github-actions[bot]' &&
      item.body?.startsWith('## Todo Commenter\n')
    );
  });

  if (comment) {
    return await octokit.rest.pulls.updateReviewComment({
      owner,
      repo,
      body,
      comment_id: comment.id,
      pull_number: prNumber
    });
  }

  return octokit.rest.pulls.createReview({
    owner,
    repo,
    event: block ? 'REQUEST_CHANGES' : 'COMMENT',
    pull_number: prNumber,
    body
  });
}
