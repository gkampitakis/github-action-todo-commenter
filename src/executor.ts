import { context } from '@actions/github';
import { readdir } from 'fs';

export function executor(tags: string[]) {
  const testFolder = './';

  // We need to do the checkout to that branch
  // Take the diff, read each file and find occurrences of tags

  readdir(testFolder, (err, files) => {
    files.forEach(file => {
      console.log(file);
    });
  });

  console.log(context.payload, tags);
}
