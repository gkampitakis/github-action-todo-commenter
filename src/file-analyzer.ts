import { createReadStream } from 'fs';
import { createInterface } from 'readline';
import path from 'path';

type FileAnalyzerResult = {
  file: string;
  comments: Comments;
}
type Comments = Record<string, string[]>;
type EnhancedTag = { tag: string, regex: RegExp };

export async function fileAnalyzer(filePaths: string[], tags: string[]): Promise<FileAnalyzerResult[]> {
  const enhancedTags = tags.map(tag => ({ tag, regex: new RegExp(tag + '(.*)') }));
  const ignoreMinSize = Math.min(...tags.map(t => t.length));
  const result: FileAnalyzerResult[] = [];

  for (const filePath of filePaths) {
    const absolutePath = path.join(process.cwd(), filePath);
    const comments = await readFile(absolutePath, enhancedTags, ignoreMinSize);

    result.push({
      file: filePath,
      comments
    });
  }

  return result;
}

async function readFile(filepath: string, tags: EnhancedTag[], ignoreMinSize: number): Promise<Comments> {
  const comments: Comments = {};
  const stream = createReadStream(filepath, 'utf-8');
  const rl = createInterface({
    input: stream,
    crlfDelay: Infinity
  });

  for await (const line of rl) {
    if (ignoreMinSize > line.length) continue;

    for (let { tag, regex } of tags) {
      const matched = line.match(regex);

      if (!matched) continue;

      const comment = matched[1].trim();

      if (comments[tag]) {
        comments[tag].push(comment);
        continue;
      }

      comments[tag] = [comment];
    }
  }

  return comments;
}
