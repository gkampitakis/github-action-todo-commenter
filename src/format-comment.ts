import { FileAnalyzerResults } from './types';

export function formatComment(
  analyzedCommentsPerFile: FileAnalyzerResults,
  { actor, reviewMsg }: { actor?: string; reviewMsg?: string }
) {
  let message = '## Todo Commenter\n';

  analyzedCommentsPerFile.forEach(({ comments, file }) => {
    message += `<details>\n`;
    message += `<summary><strong>${file}</strong></summary>\n`;

    Object.entries(comments).forEach(([key, items]) => {
      message += `\n##### ${key}\n\n`;

      items.forEach(item => {
        if (item === '') {
          message += `- [ ] <space>\n`;
          return;
        }
        message += `- [ ] ${item}\n`;
      });
    });

    message += `</details>\n\n`;
  });

  if (reviewMsg) {
    message += '---\n';
    message += `@${actor} ${reviewMsg}`;
  }

  return message;
}
