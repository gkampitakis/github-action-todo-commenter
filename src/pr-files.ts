import { GetFilesParams } from './types';
import { exec as cbExec } from 'child_process';
import { promisify } from 'util';
import { warning } from '@actions/core';

const exec = promisify(cbExec);

export async function getFiles({
  octokit,
  owner,
  repo,
  prNumber,
  base,
  head,
  ignoreFilesPattern
}: GetFilesParams) {
  const { data: prFiles } = await octokit.rest.pulls.listFiles({
    owner,
    repo,
    pull_number: prNumber,
    mediaType: {
      format: 'diff'
    }
  });

  // Filter out submodules
  const untracked = ['removed', 'unchanged'];
  let matcher = (file: typeof prFiles[number]) =>
    !untracked.includes(file.status) &&
    Boolean(file.additions) &&
    Boolean(file.blob_url);

  if (ignoreFilesPattern) {
    const regex = new RegExp(ignoreFilesPattern);
    matcher = (file: typeof prFiles[number]) =>
      !untracked.includes(file.status) &&
      Boolean(file.blob_url) &&
      Boolean(file.additions) &&
      !file.filename.match(regex);
  }

  return Promise.all(
    prFiles.filter(matcher).map(async ({ filename, patch }) => ({
      filename,
      // Noticed on big diffs the api doesn't return a patch string
      // Could not find documentation around size limit on `patch` field
      patch: patch ? patch : await diff(base, head, filename)
    }))
  );
}

async function diff(
  base: string,
  head: string,
  filename: string
): Promise<string> {
  try {
    const { stdout, stderr } = await exec(
      `git diff origin/${base}...origin/${head} -- ${filename}`
    );
    if (stderr) {
      warning(stderr);
      return '';
    }

    return stdout;
  } catch (err: any) {
    warning(err);
    return '';
  }
}
