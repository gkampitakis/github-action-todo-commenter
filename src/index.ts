import { getInput } from '@actions/core';
import { executor } from './executor';

const tags = ['TODO', 'FIXME'];
const customTags = getInput('tags').split(',');

console.log(tags, customTags);


executor(customTags || tags);
