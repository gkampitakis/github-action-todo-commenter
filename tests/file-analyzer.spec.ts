import { fileAnalyzer } from '../src/file-analyzer';

const mockTags = ['tODo:', 'FIXME:', 'NOTE'];
const absolutePaths = ['./tests/mockFiles/mockFile0.js'];

describe('FileAnalyzer', () => {

  it('should returned all identified comments', async () => {
    const result = await fileAnalyzer(absolutePaths, mockTags);

    expect(result).toMatchSnapshot();
  });
});