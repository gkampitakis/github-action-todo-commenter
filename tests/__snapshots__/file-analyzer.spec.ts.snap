// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileAnalyzer should returned all identified comments 1`] = `
Array [
  Object {
    "comments": Object {
      "FIXME:": Array [
        ":add implementation",
      ],
      "NOTE": Array [
        "Please don't forget review",
        "another note",
      ],
      "tODo:": Array [
        "this should present",
      ],
    },
    "file": "./tests/mockFiles/mockFile0.js",
  },
]
`;
